### PR TITLE
`<future>` cleanups

### PR DESCRIPTION
**Pull request by achabense**
- replaces long trailing `decltype` return types with `decltype(auto)` in several places
- removes some redundant `_Packaged_state` constructors
- simplifies signatures of `_Packaged_state::_Get_fn`, and remove an extra copy construction around it
- makes some `using`s private in `packaged_task`
- removes some uncommon `forward` patterns

### CODE REVIEW

**Comment by StephanTLavavej** [comment]
Can you please provide PR descriptions so that reviewers, and future code archaeologists, can quickly understand what is being changed and why? Only for something completely trivial like "Fix comment typos" is a PR title sufficient.

**Comment by achabense** [comment]
Hmm, added. (Sorry for my late response, this is because I'm in a different time zone)

**Comment by frederick-vs-ja** [comment]
https://github.com/microsoft/STL/blob/ed8150e099f6124c50dd4f002cd2ab8c429a81e3/stl/inc/future#L1403

Now we should say `abort()` instead. Should we change the comment in this PR or file a "good first issue"?

**Comment by achabense** [comment]
I think the usage of `forward` looks strange here. Can we replace it with `move` or `static_cast<_Mybase&&>`?
https://github.com/microsoft/STL/blob/6c69a73911b33892919ec628c0ea5bbf0caf8a6a/stl/inc/future#L963

**Comment by CaseyCarter** [comment]
> I think the usage of `forward` looks strange here. Can we replace it with `move` or `static_cast<_Mybase&&>`?

I'm not bothered by this usage of `forward` because I've heard Howard Hinnant describe why it's an intended usage. That said, it certainly never caught on and I don't consider it idiomatic. I would much rather have `static_cast<_Mybase&&>` here than `std::move` if we're going to make a change: I like it explicit that there's something weird happening here.

File: stl/inc/future
Comment (achabense) [inline stl/inc/future:1387]: (  I find `"see GH-3888"` a bit awkward, as the reasons has been explained at the beginning of the class :|  )

File: stl/inc/future
Comment (StephanTLavavej) [inline stl/inc/future:1367]: No change requested: The difference here is that `decltype(auto)` won't SFINAE, but it appears that we don't need SFINAE here (nothing can directly ask `is_invocable`).

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I think we can merge this without a second maintainer approval.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
:hourglass_flowing_sand: :stopwatch: :broom:
===

**Pull request merged**
